### ISSUE TITLE
.netcore 3.0 Services and unity are clashing

### ISSUE THREAD

**Issue by Grompot77** [PrismLibrary/Prism#1765]
Hi guys

It has been many years since I have worked with Prism, think it was 5.0 with MEF.  With the new work around netcore 3.0, I thought it was a good time to brush up on the tech.  Instead of heading off to MVVM light or something else, my first port of call was Prism.

First I started doing just the new WPF library, using the default DI from MS, but then thought about modularity and got the latest Prism ci builds.  Individually I can wire up both, the stock standard WPF does not have ViewLocation finder (not from what I could see), then Prism lends itself well for that scenario, but when trying to use the 2 together.

> I got my head smashed in a bit. 

The idea is to host some work via the DI from MS, such as database, custom libraries and HttpClient services (I have an extensive library set already working with IHostBuilder and are extremely portable), while Prism takes care of the ViewModel and Views:

_The exact sticking point is in loading the window from OnStartup, if I remove it, then it never triggers and the CreateShell is never triggered, it just continues spooling._

The sticking point right now is 'Set property 'Prism.Regions.RegionManager.RegionName' threw an exception.' Line number '16' and line position '25'.
Inner Exception: 
Message:  ServiceLocationProvider must be set.
StackTrace:
   at CommonServiceLocator.ServiceLocator.get_Current()
   at Prism.Regions.RegionManager.CreateRegion(DependencyObject element)
   at System.Windows.DependencyObject.OnPropertyChanged(DependencyPropertyChangedEventArgs e)
   at System.Windows.FrameworkElement.OnPropertyChanged(DependencyPropertyChangedEventArgs e)
   at System.Windows.DependencyObject.NotifyPropertyChange(DependencyPropertyChangedEventArgs args)
   at System.Windows.DependencyObject.UpdateEffectiveValue(EntryIndex entryIndex, DependencyProperty dp, PropertyMetadata metadata, EffectiveValueEntry oldEntry, EffectiveValueEntry& newEntry, Boolean coerceWithDeferredReference, Boolean coerceWithCurrentValue, OperationType operationType)
   at System.Windows.DependencyObject.SetValueCommon(DependencyProperty dp, Object value, PropertyMetadata metadata, Boolean coerceWithDeferredReference, Boolean coerceWithCurrentValue, OperationType operationType, Boolean isInternal)
   at System.Windows.DependencyObject.SetValue(DependencyProperty dp, Object value)
   at MS.Internal.Xaml.Runtime.ClrObjectRuntime.SetValue(Object inst, XamlMember property, Object value)

**The Code Snippet:**

    public partial class App : PrismApplication
    {
        public IServiceProvider ServiceProvider { get; private set; }

        public IConfiguration Configuration { get; private set; }

        protected override Window CreateShell()
        {
            return Container.Resolve<MainWindow>();
        }

        protected override void RegisterTypes(IContainerRegistry containerRegistry) { }

        protected override void ConfigureModuleCatalog(IModuleCatalog moduleCatalog)
        {
            moduleCatalog.AddModule<ClientModule>();
        }

        protected override void OnStartup(StartupEventArgs e)
        {
            var builder = new ConfigurationBuilder()
             .SetBasePath(Directory.GetCurrentDirectory())
             .AddJsonFile("appsettings.json", optional: false, reloadOnChange: true);

            Configuration = builder.Build();

            var serviceCollection = new ServiceCollection();
            ConfigureServices(serviceCollection);

            ServiceProvider = serviceCollection.BuildServiceProvider();

            var mainWindow = ServiceProvider.GetRequiredService<MainWindow>();
            mainWindow.Show();
        }

        private void ConfigureServices(IServiceCollection services)
        {
            services.ConfigureSqliteDatabase(Configuration);

            services.AddScoped<MainWindowViewModel>();
            services.AddTransient<MainWindow>();
        }
    }

**Comment by brianlagunas** [comment]
This is not supported. If you wish to use a custom IoC container, you must create a proper container extension. You can look at the source code for examples of how to do this.

**Comment by Grompot77** [comment]
Thanks for the response, is there any other way to use Unity to add other artifacts to be used, such as a database context that needs to be injected.  And if it can be done, where can I find a sample, would love to inject HttpClient and DbContext implementation per module during "module registration".  Used to do things like that back when MEF was still supported.

**Comment by Grompot77** [comment]
Found a work-around, it works on the same principal as passing the IServiceProvider to the constructors in normal MS IoC to retrieve registered objects ... Just wanted the wire up of the configuration out of json settings and carrying it around "disconnected".

Landed up with:

        public ViewAViewModel(IServiceProvider serviceProvider)
        {
            var ctx = serviceProvider.GetRequiredService<LookupContext>();
            var country = ctx.Countries.FirstOrDefault(c => c.Alpha2Code=="ZA");
            
            Message = country == null ? "Not found" : $"{country?.Alpha3Code}\t{country?.Description}\t{country.DialingCode}";
        }

![image](https://user-images.githubusercontent.com/9252816/55912663-77f8ab00-5be3-11e9-88fb-7147ee1ea327.png)

All working now, will pull a branch and work on a proper IoC for it, just playing around for now.
Still think Prism rocks!

**Comment by brianlagunas** [comment]
Just so you know, you can do that with a normal container such as Unity.  No need to complicate your code base.  Either way, have fun.

**Comment by Grompot77** [comment]
I would believe so, never really got my head wrapped around the DI constructs, apart from MEF and what is being done in core.

I have been a back-end stack developer for too long, so lost touch, 5-7 years ago I did my last front-end type development (hence prism).  These days I focus very much on extending the MS back-end stack & components, such as generic based apicontrollers for modular dbcontext architecture, auth, integrations architectures and microservices.  Do very much deep diving into reflected constructs, ILGeneration and so forth.  

Been around since .net 1 beta, way (WAY) back, and the tech has moved a lot, so personally I leave the front-ends to the young guns and ensure solid back-end structures.

Was just testing the new vs2019 before approving it for business, and picked a topic to work on, to see how the product is, found some issues.

PS: I did subscribe to your Youtube channel, watched a few episodes, was interesting, thanks
